onadata/apps/logger/management/commands/update_attachment_storage_bytes.py

```diff
@@ -0,0 +1,177 @@
+from django.db import transaction
+from django.db.models import Sum
+
+from onadata.apps.logger.models.attachment import Attachment
+from onadata.apps.logger.models.instance import Instance
+from onadata.apps.logger.models.xform import XForm
+from onadata.apps.main.models.user_profile import UserProfile
+from onadata.libs.utils.jsonbfield_helper import ReplaceValues
+
+
+class Command(BaseCommand):
+    help = (
+        'Retroactively add the total of '
+        'the storage file per xform and user profile'
+    )
+
+    def handle(self, *args, **kwargs):
+        self.verbosity = kwargs['verbosity']
+
+        # Dictionary which contains the total of bytes for each user. Useful
+        # to avoid another aggregate query on attachments to calculate users
+        # global usage.
+        total_per_user = defaultdict(int)
+
+        # Release any locks on the users' profile from getting submissions
+        UserProfile.objects.exclude(
+            metadata__attachments_counting_status='complete'
+        ).update(
+            metadata=ReplaceValues(
+                'metadata',
+                updates={'submissions_suspended': False},
+            ),
+        )
+
+        # Get only xforms whose users' storage counters have not been updated yet.
+        up_queryset = UserProfile.objects.values_list('user_id', flat=True).filter(
+            metadata__attachments_counting_status='complete'
+        )
+        xforms = (
+            XForm.objects.exclude(user_id__in=up_queryset)
+            .values('pk', 'user_id', 'user__username')
+            .order_by('user_id')
+        )
+
+        last_xform = None
+        for xform in xforms:
+
+            if not last_xform or (last_xform['user_id'] != xform['user_id']):
+                # Retrieve or create user's profile.
+                (
+                    user_profile,
+                    created,
+                ) = UserProfile.objects.get_or_create(user_id=xform['user_id'])
+
+                # Some old profiles don't have metadata
+                if user_profile.metadata is None:
+                    user_profile.metadata = {}
+
+                # Set the flag to true if it was never set.
+                if not user_profile.metadata.get('submissions_suspended'):
+                    # We are using the flag `submissions_suspended` to prevent
+                    # new submissions from coming in while the
+                    # `attachment_storage_bytes` is being calculated.
+                    user_profile.metadata['submissions_suspended'] = True
+                    user_profile.save(update_fields=['metadata'])
+
+                # if `last_xform` is not none, it means that the `user_id` is
+                # different from the previous one in the loop so the user
+                # profile must be updated
+                if last_xform:
+                    self.update_user_profile(
+                        last_xform, total_per_user[last_xform['user_id']]
+                    )
```

Comment:
Exactly.

Reply:
OK, I finally understand :sweat_smile: 
Could we reorganize this by changing the outer condition:
```python
            if not last_xform or (last_xform['user_id'] != xform['user_id']):
                # Retrieve or create user's profile.
                …
```
…to:
```python
            if last_xform['user_id'] != xform['user_id']:
                if last_xform:
                    # All forms for the previous user are complete; update that user's profile
                    self.update_user_profile(
                        last_xform, total_per_user[last_xform['user_id']]
                    )

                # Retrieve or create user's profile.
                …
```
…? I think dealing with all the user-switching stuff in one place would make what's happening more obvious.

Reply:
We need to keep `if not last_xform or`.  Otherwise the iteration will raise an error when `last_xform` is `None` and we try compare `last_xform['user_id'] != xform['user_id']`.